File: problem207.py
"""

[1,0] -> 0 before 1

0->1
1->2
2->3
3->0

0->1->2->3
-
|     -
|    /
|   /
|  /
| /
4


                         4<---- 0 ----> 1
                                ^       |
                                |       |
                                |       |
                                |       |
                                |       v
                                3<----- 2

0->1
1->0

"""
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        
        def Traverse(node):
            nonlocal courses, currentpath, visited
            #print(node, currentpath, visited)
            for neighbour in courses[node]:
                if neighbour in currentpath:
                    return False
                if neighbour not in visited:
                    visited.add(neighbour)
                    currentpath.add(neighbour)
                    if not Traverse(neighbour):
                        return False
                    currentpath.remove(neighbour)
            return True
                 
        courses = defaultdict(set)
        currentpath = set()
        visited= set()
        for i in range(numCourses):
            courses[i]
            
        for p in prerequisites:
            courses[p[1]].add(p[0])
        
        #print(courses, courses.keys())
        
        for c in courses.keys():
            if c not in visited:
                visited.add(c)
                currentpath.add(c)
                if not Traverse(c):
                    return False
                currentpath.remove(c)
                
        return True   
        

File: problem207.cs
/*

0   ->  1
1   ->  2
2   ->  3
3   ->  0


*/


public class Solution {
    Dictionary<int, List<int>> adjacencyMatrix;
    HashSet<int> visited, path, checkedNode;
    public bool CanFinish(int numCourses, int[][] prerequisites)
    {
        adjacencyMatrix= new Dictionary<int, List<int>>();
        checkedNode = new HashSet<int>();
        
        foreach(int[] p in prerequisites)
        {
            if(!adjacencyMatrix.ContainsKey(p[1]))
                adjacencyMatrix[p[1]] = new List<int>();
            if(!adjacencyMatrix.ContainsKey(p[0]))
                adjacencyMatrix[p[0]] = new List<int>();
            adjacencyMatrix[p[1]].Add(p[0]);
        }
        
        foreach(var kvp in adjacencyMatrix)
        {
            path = new HashSet<int>();
            path.Add(kvp.Key);
            if(!Dfs(kvp.Key))
            {
                return false;
            }
            path.Remove(kvp.Key);
        }
        return true;
    }
    
    public bool Dfs(int node)
    {
        if(checkedNode.Contains(node))
            return true;
        
        foreach(int next in adjacencyMatrix[node])
        {
            if(path.Contains(next))
                return false;
            path.Add(next);
            if(!Dfs(next))
                return false;
            path.Remove(next);
        }
        checkedNode.Add(node);
        return true;
    }
}

//Time: O(N)
//space: O(N)
